File: examples/cpp/ip_modification/ip_modification.cpp
#include <chrono>
#include <iostream>
#include <thread>
#include <opendaq/opendaq.h>

void retrieveIpConfiguration(const daq::NetworkInterfacePtr& iface)
{
    try
    {
        auto config = iface.requestCurrentConfiguration();
        std::cout << "Parameters:"
                  << "\n- dhcp4: " << config.getPropertyValue("dhcp4")
                  << "\n- address4: " << config.getPropertyValue("address4")
                  << "\n- gateway4: " << config.getPropertyValue("gateway4")
                  << "\n- dhcp6: " << config.getPropertyValue("dhcp6")
                  << "\n- address6: " << config.getPropertyValue("address6")
                  << "\n- gateway6: " << config.getPropertyValue("gateway6")
                  << std::endl;
    }
    catch (const std::exception& e)
    {
        std::cout << "Error: " << e.what() << std::endl;
    }
}

int main(int /*argc*/, const char* /*argv*/[])
{
    using namespace daq;

    // Create an Instance, loading modules at MODULE_PATH
    const InstancePtr instance = Instance(MODULE_PATH);

    auto availableDevices = instance.getAvailableDevices();

    for (const auto& deviceInfo : availableDevices)
    {
        if (deviceInfo.getName() == "Reference device simulator")
        {
            std::cout << "Found simulator device: " << deviceInfo.getConnectionString() << std::endl;
            std::cout << "Retrieving old IP config..." << std::endl;
            retrieveIpConfiguration(deviceInfo.getNetworkInterface("enp0s3"));

            auto config = deviceInfo.getNetworkInterface("enp0s3").createDefaultConfiguration();
            config.setPropertyValue("dhcp4", False);
            config.setPropertyValue("address4", String("192.168.56.155/24"));
            config.setPropertyValue("gateway4", String("192.168.56.1"));

            try
            {
                std::cout << "Device " << deviceInfo.getConnectionString() << " submitting new IP config..." << std::endl;
                deviceInfo.getNetworkInterface("enp0s3").submitConfiguration(config);
            }
            catch (const std::exception& e)
            {
                std::cout << "Error: " << e.what() << std::endl;
                return 0;
            }

            std::cout << "Wait 5 seconds for new configuration applied ..." << std::endl;
            std::this_thread::sleep_for(std::chrono::seconds(5));

            std::cout << "Retrieving updated IP config..." << std::endl;
            retrieveIpConfiguration(deviceInfo.getNetworkInterface("enp0s3"));

            return 0;
        }
    }

    std::cout << "No simulator device found" << std::endl;
    return 0;
}

File: examples/python/ip_modification.py
import sys
import opendaq

def change_ip_config(device_info: opendaq.IDeviceInfo):
    net_interface = device_info.get_network_interface('enp0s3')

    dhcp4 = False
    addresses4 = opendaq.List()
    addresses4.push_back(opendaq.String('192.168.56.155/24'))
    gateway4 = opendaq.String('192.168.56.1')

    config = net_interface.create_default_configuration()
    config.set_property_value('dhcp4', dhcp4)
    config.set_property_value('addresses4', addresses4)
    config.set_property_value('gateway4', gateway4)

    try:
        print('Submitting new IP configuration parameters ...')
        net_interface.submit_configuration(config)
        print('Done')
    except Exception as e:
        print(f'Failed: {str(e)}', file=sys.stderr)

def main():
    # Create a fresh openDAQ(TM) instance that we will use for all the interactions with the openDAQ(TM) SDK
    instance = opendaq.Instance()

    # Find and connect to a simulator device
    for device_info in instance.available_devices:
        if device_info.name == 'Reference device simulator':
            change_ip_config(device_info)

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass

